### DISCUSSION TITLE
Adding Route Functionality to flutter-maplibre-gl

### DISCUSSION THREAD

**Discussion by hatch01** [maplibre/flutter-maplibre-gl#242] [Ideas]
Hello everyone!

I was wondering if there are any plans to add route functionality to flutter-maplibre-gl. Routing is an essential feature in many location-based applications, and having this capability within the library would be incredibly helpful. It would allow us to easily display routes, calculate directions, and provide turn-by-turn navigation to our users.

I understand that there are alternative solutions available, such as using external routing APIs or integrating other Flutter packages specifically designed for routing. Nonetheless, having route functionality directly within flutter-maplibre-gl would streamline the development process and provide a more cohesive experience.

Thank you all for your contributions and support in advance. I look forward to your insights and suggestions!
Best regards,

**Comment by mariusvn** [comment]
I agree that it would be interesting.

The repo may still needs some stability work before implementing it tbh but it wouls be really nice to have, yes.

**Comment by ishafiul** [comment]
I think it's time MapLibre flutter had navigation. MapLibre has iOS and Android SDKs, so why not use it with Flutter? I'm very interested in working on this, but I think I need some guidelines.

**Reply by josxha** [reply]
I haven't looked into navigation with MapLibre myself yet. Because the iOS and Android navigation are implemented in their own SDKs I think it would be best to do the same for flutter too? I can try to help out with questions where I can.

There is also a new project called Ferrostar developed by StadiaMaps and written in Rust. I think it's still getting implemented but [looking at their README](https://github.com/stadiamaps/ferrostar?tab=readme-ov-file#current-status ) they have now support for iOS and Android. Because it is built in Rust it could use the [flutter_rust_bridge](https://cjycode.com/flutter_rust_bridge/) that has extensive code generation.
Maybe @ianthetechie could give some insights about the current state of the project and if they already considered bindings for flutter?

**Reply by ishafiul** [reply]
I also think that creating different SDKs for navigation would make more sense. I'll also join Ferrostar's Slack to better understand which one will be the best.

**Reply by ianthetechie** [reply]
Yes! Thanks for mentioning me @josxha!

> Because the iOS and Android navigation are implemented in their own SDKs I think it would be best to do the same for flutter too? I can try to help out with questions where I can.

Yeah, same opinion here. Specifically (also addressing the OP @hatch01's remarks), I think that **routing** (data structure describing a sequence of maneuvers; API or local using something like [valhalla mobile](https://github.com/Rallista/valhalla-mobile)), **navigation** (real-time guidance through the steps, notification of deviation, recalculation, etc.), and **map rendering** are all separate. I think they should be separate at least at the library level, but that they can be composable to some extent. Let me expand on that a bit...

> There is also a new project called Ferrostar developed by StadiaMaps and written in Rust. I think it's still getting implemented but [looking at their README](https://github.com/stadiamaps/ferrostar?tab=readme-ov-file#current-status) they have now support for iOS and Android. Because it is built in Rust it could use the [flutter_rust_bridge](https://cjycode.com/flutter_rust_bridge/) that has extensive code generation.

Sorry for the long prelude, but that's what you need to understand the approach Ferrostar takes. It provides **shared data models** for core concepts like routes, offers an extensible **"adapter" interface for working with routing engines** (you can define these arbitrarily in user code as well without waiting for a PR), **composable UI components** which are platform-dependent (SwiftUI, Jetpack Compose, etc.), and **platform-dependent maps**.

The core (state management etc.) is more or less feature-complete. The (native) iOS and Android platforms have functional UIs which are solid beta and slightly rougher beta quality respectively.

I would also very much like to support Flutter (and React Native) as distinct platforms too. To make this happen, we'd need to wrap the (Rust) core and create a composable, idiomatic UI. The UI would ideally use Flutter MapLibre GL. You're not the first person to ask me about this recently, so there's an issue for the [core](https://github.com/stadiamaps/ferrostar/issues/16) and [UI](https://github.com/stadiamaps/ferrostar/issues/106). Feedback + contribution welcome :) Let me know as I'm actually talking with a company this week that might contribute dev resources to make it happen so I want to coordinate efforts among anyone interested.
